builtins.go

```diff
@@ -371,6 +376,85 @@ func builtinFilter(i *interpreter, trace TraceElement, funcv, arrv value) (value
 	return makeValueArray(elems), nil
 }
 
+type sortData struct {
+	i      *interpreter
+	trace  TraceElement
+	values []value
+	keys   []value
+	err    error
+}
+
+func (d *sortData) Len() int {
+	return len(d.values)
+}
+
+func (d *sortData) Less(i, j int) bool {
+	b, err := valueLess(d.i, d.trace, d.keys[i], d.keys[j])
+	if err != nil {
+		d.err = err
+		panic("Error while comparing elements")
+	}
+	return b
+}
+
+func (d *sortData) Swap(i, j int) {
+	d.values[i], d.values[j] = d.values[j], d.values[i]
+	d.keys[i], d.keys[j] = d.keys[j], d.keys[i]
+}
+
+func (d *sortData) Sort() (err error) {
+	defer func() {
+		if d.err != nil {
+			if r := recover(); r != nil {
+				err = d.err
+			}
+		}
+	}()
+	sort.Stable(d)
+	return
+}
+
+func arrayFromValues(vs []value) *valueArray {
+	thunks := make([]*cachedThunk, len(vs))
+	for i := range vs {
+		thunks[i] = readyThunk(vs[i])
+	}
+	return makeValueArray(thunks)
+}
+
+func builtinSort(i *interpreter, trace TraceElement, arguments []value) (value, error) {
+	arrv := arguments[0]
+	keyFv := arguments[1]
+
+	arr, err := i.getArray(arrv, trace)
+	if err != nil {
+		return nil, err
+	}
+	keyF, err := i.getFunction(keyFv, trace)
+	if err != nil {
+		return nil, err
+	}
+	num := arr.length()
+
+	data := sortData{i: i, trace: trace, values: make([]value, num), keys: make([]value, num)}
+
+	for counter := 0; counter < num; counter++ {
+		var err error
+		data.values[counter], err = arr.elements[counter].getValue(i, trace)
```

Comment:
We need to document a minor change in behavior in the release:
```jsonnet
std.length(std.sort([error 'lol', error 'lol2'], function(key)0))
```
Will now error out as the thunks are always evaluated.

Reply:
Hmmm... perhaps there is no reason to evaluate them, since these are only swapped and never actually compared.

Generally though, I think that what parts of arguments are evaluated should be considered an implementation detail which may change freely from version to version, unless the documentation is very specific (which it isn't in this case). Even minor changes to the algorithm, data validation, little optimizations can change what exactly gets evaluated. And it's a bit tricky to test, too.

Reply:
It'd be easier to test if there was in general a way to test for errors.  But that's not coming any time soon.

Reply:
Fixed:
```
$ ./jsonnet -e "std.length(std.sort([error 'lol', error 'lol2'], function(key)0))"
2
```